File: c++/fizzbuzz.cpp
#include <iostream>
using namespace std;

template<bool T>
struct Fizz {
    Fizz () { cout << "Fizz"; }
};

template<>
struct Fizz<false> { };

template<bool T>
struct Buzz {
    Buzz () { cout << "Buzz"; }
};

template<>
struct Buzz<false> { };

template<bool T, int N>
struct Terminate {
    Terminate () { cout << endl; }
};

template<int N>
struct Terminate<false, N> {
    Terminate () { cout << N << endl; }
};

template<int N>
struct FizzBuzz {
    FizzBuzz () {
	const static bool 
	    fizz = N % 3 == 0,
	    buzz = N % 5 == 0;
	Fizz<fizz>(); 
	Buzz<buzz>();
	Terminate<fizz || buzz, N>();
	FizzBuzz<N+1>();
    }
};

template<>
struct FizzBuzz<101> { };

int main()
{
    FizzBuzz<1>();
    return 0;
}

File: haskell/fizzbuzz.hs
data FizzBuzz a = FizzBuzz (String, a)

instance (Show a) => Show (FizzBuzz a) where
    show (FizzBuzz ("",x)) = show x
    show (FizzBuzz (m, x)) = m

instance Monad FizzBuzz where
    return x = FizzBuzz ("", x)
    p >>= q = FizzBuzz (m ++ m' , x')
        where
          FizzBuzz (m, x) = p
          FizzBuzz (m', x') = q x

rule n s = (\x -> FizzBuzz $ if (x `mod` n) == 0 then (s, x) else ("", x))

fizzbuzz = map (\x -> return x >>= (rule 3 "Fizz") >>= (rule 5 "Buzz"))
